### PR TITLE
support DATETRUNC() function

### PR DESCRIPTION
**Pull request by skumawat2025**
### Description

With this pull request babelfish now supports function DATETRUNC(). 
 
The DATETRUNC function returns an input date truncated to a specified datepart.

For the implementation of the 'DATETRUNC' function, the PostgreSQL 'date_trunc' function was utilized with certain custom modifications to meet specific requirements.

#### Example - 

```
SELECT DATETRUNC(year, CAST('2021-12-08 11:30:15.1234567' as datetime2))
2021-01-01 00:00:00.0000000

SELECT DATETRUNC(month, '1998-03-04');
1998-03-01 00:00:00.0000000
```
For detailed explanation - 
Reference - https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/t-sql/functions/datetrunc-transact-sql.md

#### Limitation 
There are some limitation of current implementation of DATETRUNC() function, when a specific scale provided during casting to time, datetime2, datetimeoffset datatypes. 
1.  The output should match the fractional time scale same as the input data type. But currently BBF output will always be till 7 fractional digits not matter what is input data type fractional time-scale. 

##### Example. 
```
SELECT DATETRUNC(MILLISECOND, CAST('09:23:32.23456' as time(4)));
EXPECTED:- 09:23:32.2340
BBF:- 09:23:32.2340000

SELECT DATETRUNC(Month, CAST('2012-02-02 09:23:32.23456' as datetime2(1)));
EXPECTED:- 2012-02-01 00:00:00.0
BBF:- 2012-02-01 00:00:00.0000000
```

2.  When the fractional time-scale of input data type is not sufficient for provided datepart,  function DATETRUNC should throw an error message, but currently BBF's definition of DATETRUNC will give output by assuming required fractional seconds digits as 0. 

##### Example. 

```
SELECT DATETRUNC(MILLISECOND, CAST('2012-02-02 09:23:32.23456' as datetime2(2)));
EXPECTED:- The datepart millisecond is not supported by date function datetrunc for data type datetime2
BBF:-2012-02-02 09:23:32.2300000

SELECT DATETRUNC(MILLISECOND, CAST('2012-02-02 09:23:32.23456 +02:12' as datetimeoffset(2)));
EXPECTED:- The datepart millisecond is not supported by date function datetrunc for data type datetimeoffset
BBF:- 2012-02-02 09:23:32.2300000 +02:12
```

Signed-off-by: Sandeep Kumawat [skumwt@amazon.com](mailto:skumwt@amazon.com)

### Issues Resolved

BABEL-3953

### Test Scenarios Covered ###
* **Use case based -** Added

* **Boundary conditions** - Added

* **Arbitrary inputs** - Added

* **Negative test cases** - Added

* **Minor version upgrade tests**- Added

* **Major version upgrade tests** - Added

* **Performance tests** - NA

* **Tooling impact** - NA

* **Client tests** - NA

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 1]
```diff
@@ -4333,3 +4333,109 @@ BEGIN
...
+CREATE OR REPLACE FUNCTION sys.DATETRUNC(IN datepart PG_CATALOG.TEXT, IN date ANYELEMENT) RETURNS ANYELEMENT AS
+$body$
+DECLARE
+    days_offset INT;
+    v_day INT;
+    result_date timestamp;
+    input_expr_timestamp timestamp;
+    date_arg_datatype regtype;
+    offset_string TEXT;
+    datefirst_value INT;
```
Comment (hash-16) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql:4431]: Please mention the schema also from which text datatype is being used.

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 2]
```diff
@@ -4333,3 +4333,109 @@ BEGIN
...
+        ELSIF date_arg_datatype = 'time'::regtype THEN
+            IF datepart IN ('year', 'quarter', 'month', 'doy', 'day', 'week', 'tsql_week') THEN
+                RAISE EXCEPTION 'The datepart ''%'' is not supported by date function datetrunc for data type ''time''.', datepart;
+            END IF;
+            -- Limitation in determining if the specified fractional scale (if provided any) for time datatype is 
+            -- sufficient for provided datepart (millisecond, microsecond) value
+        ELSIF date_arg_datatype IN ('datetime2'::regtype, 'datetimeoffset'::regtype) THEN
+            -- Limitation in determining if the specified fractional scale (if provided any) for the above datatype
+            -- is sufficient for provided datepart (millisecond, microsecond) value;  
+        END IF;
```
Comment (hash-16) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql:4461]: here it should be insufficient in place of sufficient . Also are we not returning any thing for this case ?

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.3.0--3.4.0.sql [hunk 3]
```diff
@@ -657,6 +657,112 @@ $BODY$
...
+            RAISE EXCEPTION '''%'' is not a recognized datetrunc option.', datepart;
+    END IF;
+    BEGIN
+    input_expr_datetime2 := cast(date as sys.datetime2);
+    exception
+        WHEN others THEN
+                RAISE USING MESSAGE := 'Conversion failed when converting date and/or time from character string.';
+    END;
+    IF input_expr_datetime2 IS NULL THEN
+        RETURN NULL;
```
Comment (hash-16) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.3.0--3.4.0.sql:890]: Is this condition necessary and if yes then for which testcase ?

File: test/JDBC/input/functions/BABEL-3953-datetrunc-vu-prepare.sql [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    select 
+        datetrunc(year, cast('2020-04-15' as date)) as dt1,
+        datetrunc(yy, cast('2020-04-15' as date)) as dt2,
+        datetrunc(yyyy, cast('2020-04-15' as date)) as dt3,
+        datetrunc(quarter, cast('2020-04-15' as date)) as dt4,
+        datetrunc(qq, cast('2020-04-15' as date)) as dt5,
+        datetrunc(q, cast('2020-04-15' as date)) as dt6,
+        datetrunc(month, cast('2020-04-15' as date)) as dt7,
+        datetrunc(mm, cast('2020-04-15' as date)) as dt8,
+        datetrunc(m, cast('2020-04-15' as date)) as dt9,
```
Comment (hash-16) [inline test/JDBC/input/functions/BABEL-3953-datetrunc-vu-prepare.sql:12]: Please add all invalid testcase like datetrunc(NULL,NULL),datetrunc('NULL',NULL) etc. and others with different data types.

File: test/JDBC/input/functions/BABEL-3953-datetrunc-vu-prepare.sql [hunk 5]
```diff
@@ -0,0 +1,233 @@
+-- Test with date datatype
+CREATE VIEW DATETRUNC_vu_prepare_v1 AS (
+    select 
+        datetrunc(year, cast('2020-04-15' as date)) as dt1,
```
Comment (hash-16) [inline test/JDBC/input/functions/BABEL-3953-datetrunc-vu-prepare.sql:4]: The testcases should be verified with TSQL.

**Review by hash-16 — COMMENTED** [review]
You can add limitation of this function in PR description with some examples.

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.3.0--3.4.0.sql (hunk 3 above)
Comment (skumawat2025) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.3.0--3.4.0.sql:890]: yes, this is important when provided date expression is NULL. The output should be NULL.

**Review by hash-16 — COMMENTED** [review]
Please resolve the merge conflicts.

File: test/JDBC/input/functions/BABEL-3953-datetrunc-vu-verify.sql [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+GO
+select datetrunc('NULL', null) as dt1
+GO
+select datetrunc('NULL', 'NULL') as dt1
+GO
+select datetrunc(year, null) as dt3
+GO
+select datetrunc(years, null) as dt4
+GO
+select datetrunc(nanosecond ,null) as dt5
```
Comment (hash-16) [inline test/JDBC/input/functions/BABEL-3953-datetrunc-vu-verify.sql:77]: Add the testcase such as datetrunc('year',CAST('2020-01-01' as date)) and similar ones if not added already.

**Review by hash-16 — COMMENTED** [review]
Lets add all testcases from the doc and verify it against standard TSQL.

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 7]
```diff
@@ -4418,3 +4418,109 @@ BEGIN
...
+                RETURN concat(result_date, ' ', offset_string)::sys.datetimeoffset;
+            ELSE
+                RETURN result_date;
+            END IF;
+        END IF;
+    END;
+END;
+$body$
+LANGUAGE plpgsql STABLE;
+
```
Comment (hash-16) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql:4501]: Is this function guaranteed to return the same results given the same arguments always. ?

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 8]
```diff
@@ -4418,3 +4418,109 @@ BEGIN
...
+        date_arg_datatype := pg_typeof(date);
+        IF datepart NOT IN ('year', 'quarter', 'month', 'week', 'tsql_week', 'hour', 'minute', 'second', 'millisecond', 'microsecond', 
+                            'doy', 'day', 'nanosecond', 'tzoffset') THEN
+            RAISE EXCEPTION '''%'' is not a recognized datetrunc option.', datepart;
+        ELSIF date_arg_datatype NOT IN ('date'::regtype, 'time'::regtype, 'sys.datetime'::regtype, 'sys.datetime2'::regtype,
+                                        'sys.datetimeoffset'::regtype, 'sys.smalldatetime'::regtype) THEN
+            RAISE EXCEPTION 'Argument data type ''%'' is invalid for argument 2 of datetrunc function.', date_arg_datatype;
+        ELSIF datepart IN ('nanosecond', 'tzoffset') THEN
+            RAISE EXCEPTION 'The datepart ''%'' is not supported by date function datetrunc for data type ''%''.',datepart, date_arg_datatype;
+        ELSIF datepart IN ('dow') THEN
```
Comment (hash-16) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql:4444]: Can the  if cases for nanosecond and dow be combined ?

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 9]
```diff
@@ -4418,3 +4418,109 @@ BEGIN
...
+        ELSIF date_arg_datatype = 'smalldatetime'::regtype AND datepart IN ('millisecond', 'microsecond') THEN
+            RAISE EXCEPTION 'The datepart ''%'' is not supported by date function datetrunc for data type ''smalldatetime''.', datepart;
+        ELSIF date_arg_datatype = 'time'::regtype THEN
+            IF datepart IN ('year', 'quarter', 'month', 'doy', 'day', 'week', 'tsql_week') THEN
+                RAISE EXCEPTION 'The datepart ''%'' is not supported by date function datetrunc for data type ''time''.', datepart;
+            END IF;
+            -- Limitation in determining if the specified fractional scale (if provided any) for time datatype is 
+            -- insufficient to support provided datepart (millisecond, microsecond) value
+        ELSIF date_arg_datatype IN ('datetime2'::regtype, 'datetimeoffset'::regtype) THEN
+            -- Limitation in determining if the specified fractional scale (if provided any) for the above datatype is
```
Comment (hash-16) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql:4459]: So are we returning  nothing for this cases ?

**Review by basasairohan — APPROVED** [review]

**Review by hash-16 — APPROVED** [review]

**Review by thephantomthief — APPROVED** [review]

**Review by jsudrik — APPROVED** [review]
Approved.

**Comment by jsudrik** [comment]
With this change, babelfish now supports function DATETRUNC().

The DATETRUNC function returns an input date truncated to a specified datepart.

**Pull request merged**